File: WeAre.Athenaeum.React.Common/src/models/PageRoute.ts
import BasePageParameters from "./BasePageParameters";
import {Utility} from "@weare/athenaeum-toolkit";

export default class PageRoute {

    /**
     * Name of the route. Unlocalized.
     */
    public name: string;

    public index: number | null;

    /**
     * Id of the route.
     */
    public id: string | null;

    /**
     * Parameters of the route.
     */
    public parameters: BasePageParameters | null;

    public isPageRoute: true = true;

    constructor(name: string, index: number | null = null, id: string | null = null, parameters: BasePageParameters | null = null) {
        this.name = name;
        this.index = index;
        this.id = id;
        this.parameters = parameters;
        this.isPageRoute = true;
    }

    public static isEqual(x: PageRoute | null, y: PageRoute | null): boolean {

        if (x === y) {
            return true;
        }
        if ((x == null) && (y == null)) {
            return true;
        }
        if ((x == null) || (y == null)) {
            return false;
        }
        if (x.name !== y.name) {
            return false;
        }
        if (x.index !== y.index) {
            return false;
        }
        if (x.id !== y.id) {
            return false;
        }
        if ((x.parameters == null) && (y.parameters == null)) {
            return true;
        }

        if (Utility.getHashCode(x.parameters) != Utility.getHashCode(y.parameters)) {
            return false;
        }

        return true;
    }
}

File: WeAre.ReApptor.Common/Models/PageRoute.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.Diagnostics;
using Newtonsoft.Json;
using WeAre.ReApptor.Toolkit;

namespace WeAre.ReApptor.Common.Models
{
    [DebuggerDisplay("{Name}")]
    public class PageRoute
    {
        public string Name { get; set; }
        
        public int? Index { get; set; }
        
        public string Id { get; set; }
        
        public object Parameters { get; set; }
        
        public bool IsPageRoute => true;

        public PageRoute()
        {
        }

        public PageRoute(string name, int? index = null, string id = null, object parameters = null)
        {
            Name = name;
            Index = index;
            Id = id;
            Parameters = parameters;
        }

        public static implicit operator PageRoute(string name)
        {
            return new PageRoute { Name = name };
        }
        
        private static bool IsEqual(object x, object y)
        {
            if (ReferenceEquals(x, y))
            {
                return true;
            }
            
            if ((ReferenceEquals(x, null) || (ReferenceEquals(y, null))))
            {
                return false;
            }
            
            return (JsonConvert.SerializeObject(x) == JsonConvert.SerializeObject(y));
        }

        public static bool IsEqual(PageRoute x, PageRoute y)
        {
            if (ReferenceEquals(x, y))
            {
                return true;
            }

            if (ReferenceEquals(x, null) || ReferenceEquals(y, null))
            {
                return false;
            }

            bool equals = (x.Index == y.Index);

            equals &= Utility.IsEqual(x.Name, y.Name);

            equals &= Utility.IsEqual(x.Id, y.Id, StringComparison.InvariantCulture);

            equals &= (
                ((ReferenceEquals(x.Parameters, y.Parameters))) ||
                ((!ReferenceEquals(x.Parameters, null)) && (!ReferenceEquals(y.Parameters, null)) && (JsonConvert.SerializeObject(x) == JsonConvert.SerializeObject(y)))
            );

            return equals;
        }
    }
}